### ISSUE TITLE
accept custom geometry strings when `alevin` is used as aligner

### ISSUE THREAD

**Issue by suhrig** [nf-core/scrnaseq#218]
### Description of feature

In the current workflow version (2.2.0), when the protocol `10XV1` is selected and `alevin` as aligner, the workflow fails in the step `NFCORE_SCRNASEQ:SCRNASEQ:SCRNASEQ_ALEVIN:SIMPLEAF_QUANT` with the error message `Could not succesfully parse geometry description 10xv1`. This is because `simpleaf` does not support `10xv1` geometry; only v2 and v3 are supported. It was explained to me in the Slack channel that v1 chemistry is not supported, because it requires three sets of FastQ files, which the workflow is not designed for.

`simpleaf` not only accepts predefined geometries as an argument to the `--chemistry` parameter, it also accepts user-defined geometries (see https://simpleaf.readthedocs.io/en/latest/quant-command.html). As a workaround, one could concatenate the reads from the FastQ files containing the cell barcodes and UMIs into a single FastQ. The geometry of this concatenated FastQ file could then be passed to `simpleaf`, thus enabling the processing of v1 chemistry (and possibly other chemistry). For example, the string `1{b[14]u[10]x:}2{r:}` describes the geometry of the concatenated FastQ file for v1 chemistry.

The only remaining obstacle is getting the scrnaseq workflow to accept custom geometry strings. At the moment, the schema only allows `10XV1` (which does not work with `alevin` as explained), `10XV2`, `10XV3`, and `dropseq`. Disabling parameter validation does not work either, because other parts of the workflow expect one of the four valid keywords. So additional modifications of the workflow are required. The conclusion of my discussion with @fmalmeida on Slack was that I should open this feature request here, asking to enhance the workflow in a way to accept custom geometry strings when `alevin` is used as an aligner. This would not only be useful for 10X v1 chemistry, but also other technologies which are less common.

Many thanks in advance,
Sebastian

--

As a side note to other users faced with the same situation: The workaround I currently use is to call `simpleaf add-chemistry` in the `beforeScript`:

```
process {
   withName:'NFCORE_SCRNASEQ:SCRNASEQ:SCRNASEQ_ALEVIN:SIMPLEAF_QUANT' {
      beforeScript = 'export ALEVIN_FRY_HOME=\$NXF_SCRATCH; simpleaf add-chemistry -n 10xv1 -g "1{b[14]u[10]x:}2{r:}"'
   }
}
```

This works, but it would be nice to have a proper launch form field for custom geometry strings.

**Comment by fmalmeida** [comment]
Hello hello,
Maybe a new parameter, lets say, `--custom_geometry` could be added. Then, this would mean that, for the modules that can use a custom geometry, the pipeline would automatically set the 'preparations' necessary for it to run.

Would that be a possible solution?

One discussion would also be if this should go to the distributed module itself or only in the pipeline by using custom configs as the one you shared.

**Comment by suhrig** [comment]
> Maybe a new parameter, lets say, `--custom_geometry` could be added.

Yes, that would work. The `--chemistry` parameter should have another argument in this case, perhaps `custom`, because the predefined values would in many cases be incorrect. Alternatively, it should be possible to leave it empty.

> if this should go to the distributed module itself or only in the pipeline by using custom configs

I have no opinion on that one. I'm not familiar enough with the structure/design of the workflow to make that decision.

**Comment by fmalmeida** [comment]
Hi @suhrig ,
Sorry for delaying on it. I will start this implementation this week. Just to guarantee, the solution is, not only having a `--custom_genometry` but also having a `--custom_chemistry` right?

I could then write the module so that, either both, or none are used. If user set only geometry, then chemistry is passed as empty by default, right?

**Comment by suhrig** [comment]
What I had in mind was `--chemistry custom --custom-geometry ...`. I don't see a need for a parameter `--custom-chemistry`. Am I missing something?

**Comment by fmalmeida** [comment]
Nice, I see. I was the one missing something haha. 
Now I get the idea.

**Comment by suhrig** [comment]
Thank you for implementing this. Please give me a few days. I will give this a try as soon as I find the time.

**Comment by suhrig** [comment]
Sorry for the long delay in getting back to you. I have been busy with so many things and after finding a workaround back then didn't have any pressure to get this to work.

Anyhow, your changes almost work apart from these:

1. In the JSON file `custom_geometries.json`, the quotes get lost. The workflow creates a file with the following content:

```
{ 10xv1: 1{b[14]u[10]x:}2{r:} }
```

But there should be quotes around the strings like so:

```
{ "10xv1": "1{b[14]u[10]x:}2{r:}" }
```

2. `simpleaf` apparently ignores geometries for chemistries that it knows. When I set a custom geometry for `10xv3`, for example, it still uses what is defined internally in the software rather than what I specify. I don't think there is much we can do about this in the workflow. This would need fixing upstream.

3. It's a bit confusing that the workflow uses uppercase names for chemistries (`10XV3`), but `sampleaf` uses lowercase (`10xv3`). So in order to override, one has to specify `--custom_chemistry=10xv3`, but `--protocol=10XV3`. It's not a big issue, though. What would be best is if the user didn't have to specify `custom_chemistry` at all when one of the recognized values is used and that the pipeline deals with the upper-/lowercase mapping internally.

**Comment by grst** [comment]
@fmalmeida, could we do the same for kallisto? 
`kb` supports specifying custom technologies as described here: 
https://pachterlab.github.io/kallisto/manual#bus

> Additionally kallisto bus will accept a string specifying a new technology in the format of bc:umi:seq where each of bc,umi and seq are a triplet of integers separated by a comma, denoting the file index, start and stop of the sequence used. For example to specify the 10xV2 technology we would use 0,0,16:0,16,26:1,0,0. The first part bc is 0,0,16 indicating it is in the 0-th file (also known as the first file in plain english), the barcode starts at the 0-th bp and ends at the 16-th bp in the sequence (i.e. 16bp barcode), the UMI is similarly in the same file, right after the barcode in position 16-26 (a 10bp UMI), finally the sequence is in a separate file, starts at 0 and ends at 0 (in this case stopping at 0 means there is no limit, we use the entire sequence).

**Comment by grst** [comment]
Actually, none of this might be necessary. 

Already now, it should work to pass any technology/chemistry string via the `--protocol` parameter. If it is none of the [known ones](https://github.com/nf-core/scrnaseq/blob/ca7c05afdd5c41dc31a5a32bb207d19767836ca5/lib/WorkflowScrnaseq.groovy#L130) it will be passed on to the aligner unmodified. It ends up in the `-x` option for kallisto and the `--chemistry/-c` option for simpleaf. 
This behavior is just prevented by the paramter validation. 

The following should work (it worked for me for kallisto):

```bash
# alevin-fry, see https://simpleaf.readthedocs.io/en/latest/quant-command.html#a-note-on-the-chemistry-flag
nextflow run nf-core/scrnaseq [...] --protocol "1{b[14]u[10]x:}2{r:}" --validate_params=false --aligner=alevin
# kallisto, see https://pachterlab.github.io/kallisto/manual#bus
nextflow run nf-core/scrnaseq [...] --protocol "0,0,14:0,14,24:1,0.0" --validate_params=false --aligner=alevin
```

Instead of adding new flags, we should just fix the parameter validation and add a section in the documentation. I can take care of that.

**Comment by fmalmeida** [comment]
Awesome @grst ,
That looks like the way forward. 
🚀

**Comment by grst** [comment]
Great! 

@suhrig, would be fantastic if you could give ` --protocol "1{b[14]u[10]x:}2{r:}" --validate_params=false` a try!

**Comment by suhrig** [comment]
The pipeline runs for me even without `--validate_params=false`. The issue is that it has not the desired effect. I think what needs to be changed is that the quotes on the following line need to be double-escaped:

https://github.com/nf-core/scrnaseq/blob/a2922bb1eb5d3850ad8951ae76ea0713a92b2a33/conf/modules.config#L126

Otherwise, they are interpreted by the shell, get lost and never make it into the json file.

**Comment by grst** [comment]
@suhrig, using `--protocol "1{b[14]u[10]x:}2{r:}"` doesn't even write to that json file, but rather passes it to the `-c` flag here:
https://github.com/nf-core/scrnaseq/blob/ca7c05afdd5c41dc31a5a32bb207d19767836ca5/modules/local/simpleaf_quant.nf#L69

You still have a point about the quotes!

**Comment by grst** [comment]
as a workaround `--protocol '"1{b[14]u[10]x:}2{r:}"'` could work

**Comment by grst** [comment]
Parameter validation and quotes should be fixed in #273

**Comment by grst** [comment]
In the `dev` version, `--protocol "1{b[14]u[10]x:}2{r:}"` should work now, without any additional workarounds. 
There is no `--custom_chemistry` parameter for now, as it is already covered by `--protocol`. 

Closing this for now, if there are any issues, please let us know.
